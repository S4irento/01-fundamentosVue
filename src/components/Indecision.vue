<template>
  <img v-if="img" :src="img">
  <div class="bg-dark"></div>
  <div class="indecision-container">
    <input type="text" placeholder="Hazme una pregunta" v-model="question">
    <p>Recuerda terminar con signo de interrogación (?)</p>
    <div v-if="isValidQuestion">
      <h2>{{ question }}</h2>
      <h1>{{ answer }}</h1>
    </div>
  </div>
</template>

<script>
export default {
  data(){
    return {
      question: null,
      answer: null,
      img: null,
      isValidQuestion: false
    }
  },
  methods: {
    async getAnswer(){
      try{
        this.answer = "Procesando...";
        const {answer, image} = await fetch('https://yesno.wtf/api').then(r => r.json());
        this.answer = answer.toUpperCase();
        this.img = image;
      }catch(e){
        this.answer = "Fallo en al API";
        this.img = null;
      }
    }
  },
  watch: {
    question(value, oldValue){
      this.isValidQuestion = false;

      if(!value.includes('?')) return

      this.isValidQuestion = true;

      //TODO: realizar petición HTTP
      this.getAnswer()
    }
  }
}
</script>

<style>
  img, bg-dark{
    height: 100vh;
    left: 0px;
    max-height: 100%;
    max-width: 100%;
    position: fixed;
    top: 0%;
    width: 100vw;
  }

  .bg-dark{
    background-color: rgba(0, 0, 0, 0.514);
  }

  .indecision-container{
    position: relative;
    z-index: 99;
  }

  input{
    width: 250px;
    padding: 10px 15px;
    border-radius: 5px;
    border: none;
  }

  input:focus{
    outline: none;
  }

  p{
    color: white;
    font-size: 20px;
    margin-top: 0px;
  }

  h1, h2{
    color: white;
  }
</style>