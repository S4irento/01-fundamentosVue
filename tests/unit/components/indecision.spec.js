import { shallowMount } from "@vue/test-utils";
import Indecision from "@/components/Indecision";

describe("Indecision component", () => {
  let wrapper;

  global.fetch = jest.fn(() =>
    Promise.resolve({
      json: () =>
        Promise.resolve({
          answer: "yes",
          forced: false,
          image: "https://yesno.wtf/assets/yes/2.gif",
        }),
    })
  );

  beforeEach(() => {
    wrapper = shallowMount(Indecision);
  });

  test("debe hacer match con la SnapShot", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  test("No debe ejecutarse el getAnswer si el input no tiene un símbolo de interrogación (?)", async () => {
    const getAnswerSpy = jest.spyOn(wrapper.vm, "getAnswer");
    const input = wrapper.find("input");

    await input.setValue("¿Seré millonario");
    expect(getAnswerSpy).not.toHaveBeenCalled();
  });

  test("Debe ejecutarse el getAnswer si el input tiene un símbolo de interrogación (?)", async () => {
    const getAnswerSpy = jest.spyOn(wrapper.vm, "getAnswer");
    const input = wrapper.find("input");

    await input.setValue("¿Seré millonario?");
    expect(getAnswerSpy).toHaveReturnedTimes(1);
  });

  test("Debe validar el Fetch API", async () => {
    await wrapper.vm.getAnswer();

    const img = wrapper.find("img");

    expect(img.exists()).toBeTruthy();
    expect(wrapper.vm.img).toBe("https://yesno.wtf/assets/yes/2.gif");
    expect(wrapper.vm.answer).toBe("YES");
  });

  test("Debe validar un fallo de la fetch API", async () => {
    fetch.mockImplementationOnce(() => Promise.reject("Error API!"));

    await wrapper.vm.getAnswer();
    const img = wrapper.find('img');

    expect(img.exists()).toBeFalsy();
    expect(wrapper.vm.answer).toBe("Fallo en al API");
  })
});
