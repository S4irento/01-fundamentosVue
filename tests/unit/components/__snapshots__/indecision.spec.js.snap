// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Indecision component debe hacer match con la SnapShot 1`] = `
<!--v-if-->
<div class="bg-dark"></div>
<div class="indecision-container"><input type="text" placeholder="Hazme una pregunta">
  <p>Recuerda terminar con signo de interrogaci√≥n (?)</p>
  <!--v-if-->
</div>
`;
